File: farewell.hs
module Main (main) where

import System.Environment (getArgs)
import Text.Regex (mkRegex, subRegex, splitRegex)
import Data.Char (ord, chr)
import Data.Bits (xor)

decrypt _ _ [] decrypted = decrypted
decrypt key index encrypted decrypted = decrypt key (index + 1) (tail encrypted)
    (decrypted ++ [chr (xor (read (head encrypted) :: Int) (ord (key !! (mod index (length key)))))])

main = do
  [f] <- getArgs
  content <- readFile f
  let encrypted = splitRegex (mkRegex ",") (subRegex (mkRegex "[\n\r\t ]") content "")
  let key = "i hope in the next ten years there would be no other farewell letter brilliant than this one"
  putStr (decrypt key 0 encrypted [])

File: farewell.fs
open System
open System.IO
open System.Text.RegularExpressions

let rec decrypt (key: string) index encrypted decrypted =
    match encrypted with
    | [] -> decrypted
    | _ -> decrypt key (index + 1) (List.tail encrypted)
                   (List.append decrypted
                                [char ((int (List.head encrypted)) ^^^ (int key.[index % key.Length]))])

[<EntryPoint>]
let main (args : string[]) =
    let content = File.ReadAllText args.[0]
    let encrypted = (new Regex("\\s")).Replace(content, "").Split [|','|] |> Array.toList
    let key = "i hope in the next ten years there would be no other farewell letter brilliant than this one"
    printf "%s" (new string [|for c in (decrypt key 0 encrypted []) -> c|])
    0
